mig-blue-green/setup.sh

```diff
@@ -0,0 +1,226 @@
+echo -e "#  Managed Instance Groups, Cloud Build & Terraform  #"
+echo -e "#                                                    #"
+echo -e "######################################################${NC}\n"
+
+echo -e "\nSTARTED ${GREEN}setup.sh:${NC}"
+
+echo -e "\nIt's ${RED}safe to re-run${NC} this script to ${RED}recreate${NC} all resources.\n"
+echo "> Checking GCP CLI tool is installed"
+gcloud --version > /dev/null 2>&1
+
+readonly EXPLICIT_PROJECT_ID="$1"
+readonly EXPLICIT_CONSENT="$2"
+
+if [ -z "$EXPLICIT_PROJECT_ID" ]; then
+    echo "> No explicit project id provided, trying to infer"
+    PROJECT_ID="$(gcloud config get-value project)"
+else
+    PROJECT_ID="$EXPLICIT_PROJECT_ID"
+fi
+
+if [ -z "$PROJECT_ID" ]; then
+    echo "ERROR: GCP project id was not provided as parameter and could not be inferred"
+    exit 1
+else
+    readonly PROJECT_NUM="$(gcloud projects describe $PROJECT_ID --format='value(projectNumber)')"
+    if [ -z "$PROJECT_NUM" ]; then
+        echo "ERROR: GCP project number could not be determined"
+        exit 1
+    fi
+    echo -e "\nYou are about to:"
+    echo -e "  * modify project ${RED}${PROJECT_ID}/${PROJECT_NUM}${NC}"
+    echo -e "  * ${RED}enable${NC} various GCP APIs"
+    echo -e "  * make Cloud Build ${RED}editor${NC} of your project"
+    echo -e "  * ${RED}execute${NC} Cloud Builds and Terraform plans to create"
+    echo -e "  * ${RED}4 VMs${NC}, ${RED}3 load balancers${NC}, ${RED}3 public IP addresses${NC}"
+    echo -e "  * incur ${RED}charges${NC} in your billing account as a result\n"
+fi
+
+if [ "$EXPLICIT_CONSENT" == "yes" ]; then
+  echo "Proceeding under explicit consent"
+  readonly CONSENT="$EXPLICIT_CONSENT"
+else
+    echo -e "Enter ${BLUE}'yes'${NC} if you want to proceed:"
+    read CONSENT
+fi
+
+if [ "$CONSENT" != "yes" ]; then
+    echo -e "\nERROR: Aborted by user"
+    exit 1
+else
+    echo -e "\n......................................................"
+    echo -e "\n> Received user consent"
+fi
+
+#
+# Executes action with one randomly delayed retry.
+#
+function do_with_retry {
+    COMMAND="$@"
+    echo "Trying $COMMAND"
+    (eval $COMMAND && echo "Success on first try") || ( \
+        echo "Waiting few seconds to retry" &&
+        sleep 10 && \
+        echo "Retrying $COMMAND" && \
+        eval $COMMAND \
+    )
+}
+
+echo "> Enabling required APIs"
+# Some of these can be enabled later with Terraform, but I personally
+# prefer to do all API enablement in one place with gcloud.
+gcloud services enable \
+    --project=$PROJECT_ID \
+    cloudbuild.googleapis.com \
+    cloudresourcemanager.googleapis.com \
+    compute.googleapis.com \
+    sourcerepo.googleapis.com \
+    --no-user-output-enabled \
+    --quiet
+
+echo "> Adding Cloud Build to roles/editor"
+gcloud projects add-iam-policy-binding \
+    "$PROJECT_ID" \
+    --member="serviceAccount:$PROJECT_NUM@cloudbuild.gserviceaccount.com" \
+    --role='roles/editor' \
```

Comment:
Is it possible to identify the exact set of roles rather than use grant the editor role?

Reply:
Yes, it is possible, but I prefer we do not, here is why. The resulting list is very long with over 15 items.
Each individual role perimssion on that list are very high risk like "ability to manage GCE VMs" or "ability to control networking", etc. So ultimately I felt that giving explicit "editor" role signifies the deep trust that is required between projects.